specs/upgrade/README.md

```diff
@@ -0,0 +1,79 @@
+# 0-OS upgrade
+
+## Philosophy
+
+0-OS is meant to be a black box no one can access. While this provide some nice security features it also makes it harder to manage. Specially when it comes to update/upgrade.
+
+Since 0-OS is meant to be driven by some blockchain transaction, The version of the software running on 0-OS could also be driven by a consensus of technical people agreeing on a new version to deploy.
+
+Once a new version of a component is ready, a multisig transaction is created by the 0-OS team and X number of person needs to sign the transaction to approve the deployment of the new version.
+
+Using a blockchain a source for upgrade gives us the ability to create different version of the grid. Version matching the different network of the blockchain:
+
+- Dev: ephemeral network only setup to develop and test new features. Can be created and reset at anytime
+- Test: Mostly stable feature that needs to be tested at scale, allow preview and test of new features. Always the latest greatest. This network can be rest sometimes, by should be relatively stable.
+- Main: Released of stable version. Used to run the real grid with real money. Cannot be reset ever. Only stable and battle tested feature reach this level.
+
+Since the upgrade will be trigger by a transaction on the blockchain, building a network of node using the same version is as simple as just pointing which network of the blockchain to watch.
+
+The upgrade transaction should container a description and signature of all the new component to install. This allow to ensure the binaries installed on the node are actually a binary exact copy of what as been announced on the blockchain.
+
+0-OS could even periodically verify that all the modules it is running are not corrupted or modified by comparing the hash of the binary to the content of the blockchain.
+
+## Technical
+
+0-OS is design to provide maximum uptime for its workload, so rebooting a node should never be required to upgrade any of its component (except when we will push a kernel upgrade).
+
+The only way to get code onto a 0-OS is using flist. An upgrade flist will be composed of the new binary to install and in some case a migration binary that needs to be run before the start of the new module.
+
+The content of the upgrade flist will be cached on disk, so in the event of a power failure, the node can just restart and restart all the workloads without downloading all the new modules again.
+
+## Flist upgrade layout
+
+The files in the upgrade flist needs to be located in the filesystem tree at the same destination they would need to be in 0-OS. This allow the upgrade code to stays simple and only does a copy from the flist to the cache disk of 0-OS.
```

Comment:
Copying to the cache desk is not enough IMHO. Since the base 0-fs image is a tmpfs image that is fully loaded in memory, we need to make sure the new binaries overrides the one in the base image we can do this by one of those ways:
1- Binaries need to be copied to the `roofs` of the node, this means the files need to be copied every time the machine is boated
2- May be override the PATH variable to always prefer a (/var/cache/bin) for example over (/bin) but this can be tricky in case a library upgrade is required so we will also need LD_LIBRARY_PATH. this can get messy
3- Actually install zos on the a desk on first boot, then update the files in place.

> Talking about libraries: updating binaries without the libraries will not work (unless we grantee we always have static builds). I think instead of publishing a single flist with the updated binaries a full zos image should be published and the node then can only copy the files that need to be update (comparing checksum with local files manifest, or a tool that diff from the current 0-os image) . A node that is freshly booted can just boot from the latest build.

Reply:
> Copying to the cache desk is not enough IMHO

There is a confusion here. I used bad wording. I'm not talking about the 0-fs cache. But a disk in which 0-OS has installed what it needs to do its job.

> Binaries need to be copied to the roofs of the node, this means the files need to be copied every time the machine is boated

No I want to avoid this, all the components needs to be locally available, so if a node reboot and doesn't have internet connectivity it still can run properly

> updating binaries without the libraries will not work (unless we grantee we always have static builds). 

Of course this example only present a component that is statically build. I would like to push static binaries as much as possible, but for tools that cannot be shipped static we'll have to also distribute the libraries.

> I think instead of publishing a single flist with the updated binaries a full zos image should be published and the node then can only copy the files that need to be update (comparing checksum with local files manifest, or a tool that diff from the current 0-os image)

I don't think this is necessary except maybe for the first ever release. Then update flist can only contains the new stuff. This makes update logic so much easier if you know everything in the flist needs to be copied over.

Reply:
So does this mean zos will now require installation ?

Reply:
This is all going to be automatic of course, but yeah like today, we'll have to do some steps before the node is fully operational. There is no way around it.
These steps includes:
- reserve a disk
- do a capacity check
- make sure all modules have latest version installed
- ...